### PR TITLE
add BestEffortDebug<T>, an always-Debug wrapper over T irrespective of T: Debug / T: !Debug

### PR DESCRIPTION
**Pull request by Centril**
This adds the `BestEffortDebug` (formerly `WrapDebug`) wrapper discussed in https://github.com/rust-lang/rfcs/pull/2361 and https://github.com/rust-lang/rfcs/pull/2173.

## Motivation

This allows macros such as `debugit`, `dbg`, `p-macro` to work in generic code without `T: Debug` bounds infecting the call stack. The reason for having it in libcore is that specialization is unstable, and that `type_name` is part of `core_instrinsics` which is forever unstable. In addition, the macros in question can reuse this type and so we get some code reuse.

## Does this expose specialization?

**Yes** in the sense that you have to fall back to `type_name` and `impl<T> BEDInternal for BestEffortDebug<T>` if we decide to remove specialization entirely. The output thus becomes effectively useless, but no code will break as a result of removing specialization. The guarantees section in the documentation also explicitly states that the output format is not guaranteed.

**No** in the sense that specialization is not effectively stabilized as a result of this. The "inner helper trait" hack is used in other places such as: [`ArcFromSlice`](https://doc.rust-lang.org/nightly/src/alloc/arc.rs.html#617).

### CODE REVIEW

**Comment by scottmcm** [comment]
I do like the "there's a way to get debug output without threading `T:Debug` through" capability.

Bikeshed: `Wrap` is a super-generic word that doesn't really convey any meaning to me.  It feels like there ought to be a better word, but the only thing jumping to mind is `AlwaysDebug`.

**Comment by Centril** [comment]
# 🚲🏘️

@scottmcm  The name comes from: https://docs.rs/wrap-debug/0.1.1/wrap_debug/struct.WrapDebug.html but I'm not super excited about it as well.

`AlwaysDebug`, is kinda nice / better, but also not super exciting.

**Comment by Centril** [comment]
## More 🚲🏘️

+ `BestEffortDebug`, this one is clear about the role the type has.
+ `ValiantDebug` - a synonym of `BestEffortDebug` more or less, but probably a less known adjective.
+ `TryDebug` / `AttemptDebug` - this causes confusion with the `Try` trait and `TryFrom` (error handling in general).

Since we've been OK historically with 3-worded types such as `ExactSizeIterator` it suggests that we should be OK with `BestEffortDebug`, which is most descriptive, so I'm changing to `BestEffortDebug`.

**Comment by arthurprs** [comment]
MaybeDebug, LazyDebug, ForceDebug also come to mind.

**Comment by llogiq** [comment]
I'd like to see an extension of this that shows the ptr (either directly or hashed) to the object in question for non-Debug references. This has precedence in languages like Java, where the standard `toString()` method prints out *<type>*`@`*<address-hash>* (which is both shorter and more informative than `[<unknown> of type `*<type>*` is !Debug]`). Regarding the address, I'm personally fine with writing it out directly in hex (which means 8 digits), but others may prefer a shortened hashed version (e.g. CRC32) that is more readable while keeping the risk of accidential collisions within reasonable bounds.

**Comment by Centril** [comment]
@llogiq I specifically did not opt for such a design; the reason why is explained here: https://github.com/rust-lang/rfcs/pull/2173#issuecomment-338838834  Relevant snippet:

> We discussed including the address of `expr` in `dbg!(expr)` but since that is just the address of a temporary value inside the macro, it is actively harmful when `expr` is not a pointer. The user may also do: `dbg!(&var as *const _)` if they want pointer info.

In Java, the address is stable, so it is useful - this is not the case for the intended use case in Rust. Additionally, the output format of `BestEffortDebug<T>` is unspecified, so we can make future changes.

File: src/libcore/fmt/mod.rs [hunk 1]
```diff
@@ -1960,5 +1960,77 @@ impl<T: ?Sized + Debug> Debug for UnsafeCell<T> {
...
+
+trait BEDInternal {
+    fn fmt(&self, fmt: &mut Formatter) -> Result;
+}
+
+impl<T> BEDInternal for BestEffortDebug<T> {
+    default fn fmt(&self, fmt: &mut Formatter) -> Result {
+        use intrinsics::type_name;
+        write!(fmt, "[<unknown> of type {} is !Debug]",
+            unsafe { type_name::<T>() })
```
Comment (bluss) [inline src/libcore/fmt/mod.rs]: I think the `is !Debug` part of the message is jargon-y and hard to understand. My idea of an improvement would be `"[<unknown value> of type {}]"`

Comment (Centril) [reply src/libcore/fmt/mod.rs]: I agree on the *hard to understand* bit, how about this?
> `"[<unknown value> of type {}, which does not implement Debug]"`

The point of this being that if a library user sees this, they will know that they should bug the library author with a request to add `#[derive(Debug)]`.

**Comment by Centril** [comment]
From my discussions with @nikomatsakis (ping) on WG-traits, it seems that this particular specialization would not be permitted under [max-min specialization](http://smallcultfollowing.com/babysteps/blog/2018/02/09/maximally-minimal-specialization-always-applicable-impls/) and would require specialization predicates. Niko did say that it would be useful, but that the PR should probably be closed. So I'm not sure how to proceed...

**Comment by shepmaster** [comment]
Ping from triage, @Kimundi and @nikomatsakis ! Should we put this PR to bed, for now?

File: src/libcore/fmt/mod.rs (hunk 1 above)
Comment (bluss) [reply src/libcore/fmt/mod.rs]: (That's much better, and I think the reason for that is good)

**Comment by Centril** [comment]
An update: @aturon's [recent blog post on `specialize(..)`](http://aturon.github.io/2018/04/05/sound-specialization/) would make the following possible:

```rust
pub struct BED<T>(pub T);

impl<T> Debug for BED<T> { .. }

impl<T> Debug for BED<T> where specialize(T: Debug) { .. }
```

this would still support all types, but `: 'static` and `TyConstuctor<T, T>` would fall back to the first `impl`.
However, it would be quite uncommon to have such impls and not be generic over `'a` and `TyCtor<T, U, ..>`.

**Comment by pietroalbini** [comment]
Ping from triage! @Kimundi and @Centril, what's the status of this PR?

**Comment by aturon** [comment]
@Kimundi I think you meant:

@rfcbot fcp merge

**Comment by alexcrichton** [comment]
@rfcbot concern not-useful-if-specialization-does-not-work

Specialization is still very much in flux AFAIK and it seems like it's still not 100% clear that there's a path to stability for functionality like this. @aturon's recent blog post is indeed one possible path but I don't think we're sure that it's sound and working, right?

Additionally I do not think that functionality like this is useful in libstd if specialization is removed. That's "basically a breaking change" as the *whole point* of this type would be specialization. Otherwise it's not offering much utility over a crates.io crate that just calls `type_name` or exposing that intrinsic in a stable fashion.

This to me seems like something that should live on crates.io until we know specialization works, then it can move into libstd.

**Comment by alexcrichton** [comment]
@aturon or @Centril  do you have thoughts on the concern I listed above?

**Comment by Centril** [comment]
@alexcrichton Excuse the tardy reply ;)

I'll defer to @aturon re. the pace of implementation and stabilization of specialization.

However, you could leave `BestEffortDebug` as unstable while specialization is unstable;
you could also use `BestEffortDebug` as a way to keep `dbg!` itself unstable.
(Anyone who uses `dbg!(..)` will get an error about the unstable nature of `BestEffortDebug`...)

> Additionally I do not think that functionality like this is useful in libstd if specialization is removed.

Granted; It would become effectively useless. My biggest fear is not that the helper type itself becomes useless, but that its reverse dependency, and primary usecase, `dbg!` becomes useless as a result of it.
In this sense, there is a stronger concern ;)

> This to me seems like something that should live on crates.io until we know specialization works, then it can move into libstd.

It already does ;) Not as a standalone crate, but it does exist in the [`dbg`](https://docs.rs/dbg/1.0.4/dbg/) and [`debugit`](https://docs.rs/debugit/0.1.2/debugit/) crates. It is named `WrapDebug` and `DebugIt` in those crates.

> Otherwise it's not offering much utility over a crates.io crate that just calls `type_name` or exposing that intrinsic in a stable fashion.

Aside: Speaking of `type_name`... should there be a stable wrapper function for it?

**Comment by nikomatsakis** [comment]
I think it seems fine to have this live in crates.io and not std until such time as stabilization is on a clear path to stability (implemented in the way we would want, etc)

**Comment by alexcrichton** [comment]
Hm ok so given that specialization is not on a clear path to stabilization, @Centril are you ok holding off on this PR for now?

**Comment by Centril** [comment]
@alexcrichton Sure thing :)

**Comment by alexcrichton** [comment]
Ok!

**Pull request closed without merging**